### PR TITLE
Allow onnxruntime quantization preprocessor for dynamic quantization

### PR DESCRIPTION
**Pull request by fxmarty**
# What does this PR do?

Currently, for the onnxruntime backend, the [`QuantizationPreprocessor`](https://github.com/huggingface/optimum/blob/3b522e4d82c765660d7f60839d65ad7eda208f58/optimum/onnxruntime/preprocessors/quantization.py#L36) is usable only for static quantization to exclude nodes to quantize, because the onnx model needs to be already saved when initializing `QuantizationPreprocessor`, which was handled by [`partial_fit`](https://github.com/huggingface/optimum/blob/3b522e4d82c765660d7f60839d65ad7eda208f58/optimum/onnxruntime/quantization.py#L237) method used during calibration.

With this PR, it is possible to use `QuantizationPreprocessor` for dynamic quantization (if it happens to be relevant at some point -- at least I would like to test it), while making no change to the current workflow.

## Before submitting
- `QuantizationPreprocessor` is largely (publicly) untested and documented, in a future PR we could improve that.

This follows up https://github.com/huggingface/optimum/pull/166 , I messed up with my fork.

### CODE REVIEW

**Review by mfuntowicz — APPROVED** [review]
Thanks for taking care of this, good one!

File: optimum/onnxruntime/preprocessors/quantization.py [hunk 1]
```diff
@@ -48,11 +46,13 @@ def register_pass(self, target: PreprocessorPass):
         if target not in self._passes:
             self._passes.append(target)
 
-    def collect(self) -> Tuple[Set[str], Set[str]]:
+    def collect(self, model_or_path: Union[str, PathLike, Path, bytes]) -> Tuple[Set[str], Set[str]]:
         global_nodes_to_quantize, global_nodes_to_exclude = set(), set()
+        _graph = load_model(model_or_path.as_posix() if isinstance(model_or_path, Path) else model_or_path)
+        _model = OnnxModel(_graph)
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/onnxruntime/preprocessors/quantization.py]: My bad, I removed the `self` without the underscore!

**Comment by fxmarty** [comment]
Got a test not passing, however I think this has no link with this PR, will rerun:

```
2022-05-18T10:07:30.8335969Z ======================================================================
2022-05-18T10:07:30.8354378Z ERROR: test_dynamic_quantization (test_onnxruntime.TestORTQuantizer) (model_name='facebook/bart-base')
2022-05-18T10:07:30.8354940Z ----------------------------------------------------------------------
2022-05-18T10:07:30.8355263Z Traceback (most recent call last):
2022-05-18T10:07:30.8500187Z   File "/home/runner/work/optimum/optimum/tests/onnxruntime/test_onnxruntime.py", line 129, in test_dynamic_quantization
2022-05-18T10:07:30.8502853Z     validate_model_outputs(
2022-05-18T10:07:30.8503854Z   File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/transformers/onnx/convert.py", line 415, in validate_model_outputs
2022-05-18T10:07:30.8506013Z     raise ValueError(
2022-05-18T10:07:30.8506559Z ValueError: Outputs values doesn't match between reference model and ONNX exported model: Got max absolute difference of: 0.8644396662712097
```

from https://github.com/huggingface/optimum/blob/50cf2468903596f580102a484e50ce87b74fe4a5/tests/onnxruntime/test_onnxruntime.py#L129-L136

For me it is a little worrying that we use `atol=0.8`, why do we allow it so large @mfuntowicz  @lewtun  ? In transformers conversion, it looks like the default is 1e-5 (see https://github.com/huggingface/transformers/blob/d6b8e9cec7301ba02f642588a6f12e78ec3b9798/tests/onnx/test_onnx_v2.py#L278-L285 and https://github.com/huggingface/transformers/blob/d6b8e9cec7301ba02f642588a6f12e78ec3b9798/src/transformers/onnx/config.py#L199-L207 probably). Can the difference be so large simply due to the backend?

**Comment by fxmarty** [comment]
Rerunning, all checks are fine. Weird, not sure why there are non deterministic behavior.

**Comment by fxmarty** [comment]
Oh right, it is comparing with a static quantization model.

**Pull request merged**